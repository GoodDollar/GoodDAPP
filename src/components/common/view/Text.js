// @flow
import React from 'react'
import { StyleSheet } from 'react-native'
import { Text as PaperText } from 'react-native-paper'

import normalize from '../../../lib/utils/normalizeText'
import { withStyles } from '../../../lib/styles'
import { calculateFontFamily, calculateFontWeight } from '../../../lib/utils/fonts'

const LINE_HEIGHT_FACTOR = 1.2

class Text extends React.Component {
  _root

  /**
   * @internal
   * Required for when the component is wrapped inside a TouchableOpacity* element
   * https://facebook.github.io/react-native/docs/direct-manipulation
   */
  setNativeProps(...args) {
    return this._root && this._root.setNativeProps(...args)
  }

  render() {
    const {
      style,
      styles,
      theme,
      color,
      textAlign,
      fontWeight,
      fontFamily,
      fontSize,
      lineHeight,
      textDecorationLine,
      textTransform,
      ...rest
    } = this.props

    return <PaperText {...rest} ref={c => (this._root = c)} style={[styles.text, style]} />
  }
}

/**
 * Returns the linespacing associated to the font-size based on guidelines
 * Defaults to 1em if size is not in the list
 * @param {number} fontSize
 * @returns {*|string}
 */
const relatedLineSpacing = fontSize =>
  ({
    8: 14,
    10: 14,
    12: 16,
    14: 20,
    16: 22,
    18: 24,
    22: 30,
    24: 30,
    36: 30,
    42: 30,
  }[fontSize] || fontSize * LINE_HEIGHT_FACTOR)

const getStylesFromProps = ({
  theme,
  color,
  textAlign,
  style,
  fontWeight,
  fontFamily,
  fontSize,
  lineHeight,
  textDecorationLine,
  textTransform,
  letterSpacing,
  normalizeText = true,
}) => {
  const calculatedFontSize = Number.isFinite(fontSize) ? fontSize : 16
  const calculatedLineHeight = lineHeight || relatedLineSpacing(calculatedFontSize)

  const selectedFontFamily = theme.fonts[fontFamily] || fontFamily
  const selectedFontWeight = StyleSheet.flatten(style)?.fontWeight || fontWeight

  const calculatedFontWeight = isNaN(selectedFontWeight) ? calculateFontWeight(selectedFontWeight) : selectedFontWeight
  const calculatedFontFamily = calculateFontFamily(selectedFontFamily, selectedFontWeight)

  return {
    text: {
      color: theme.colors[color] || color || theme.colors.darkGray,
      textAlign: textAlign || 'center',
      fontWeight: calculatedFontWeight,
      fontFamily: calculatedFontFamily,
      fontSize: normalizeText ? normalize(calculatedFontSize) : calculatedFontSize,
      lineHeight: normalizeText ? normalize(calculatedLineHeight) : calculatedLineHeight,
      textTransform: textTransform || 'none',
      textDecorationLine: textDecorationLine || 'none',
      letterSpacing,
    },
  }
}

export default withStyles(getStylesFromProps)(Text)
