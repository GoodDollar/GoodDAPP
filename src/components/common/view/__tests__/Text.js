import React from 'react'
import renderer from 'react-test-renderer'
import Text from '../Text'

// Note: test renderer must be required after react-native.

describe('Text', () => {
  it('matches snapshot', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<Text>Example Text</Text>)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches snapshot with string color', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<Text color="red">Example Text</Text>)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches snapshot with theme defined color', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<Text color="primary">Example Text</Text>)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches snapshot with different textAlign', () => {
    const component = renderer.create(
      <Text color="red" textAlign="right">
        Example Text
      </Text>,
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('matches snapshot with different fontSize', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<Text fontSize={20}>Example Text</Text>)))
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
