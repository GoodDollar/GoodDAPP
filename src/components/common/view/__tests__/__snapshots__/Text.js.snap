// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text matches snapshot 1`] = `
<div
  className="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
  dir="auto"
>
  Example Text
</div>
`;

exports[`Text matches snapshot with different fontSize 1`] = `
<div
  className="css-text-901oao r-fontFamily-vnw8o6 r-fontSize-hiw28u r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
  dir="auto"
>
  Example Text
</div>
`;

exports[`Text matches snapshot with different textAlign 1`] = `
<div
  className="css-text-901oao r-color-howw7u r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-1ff274t r-textDecorationLine-13wfysu r-textTransform-3twk1y"
  dir="auto"
>
  Example Text
</div>
`;

exports[`Text matches snapshot with string color 1`] = `
<div
  className="css-text-901oao r-color-howw7u r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
  dir="auto"
>
  Example Text
</div>
`;

exports[`Text matches snapshot with theme defined color 1`] = `
<div
  className="css-text-901oao r-color-1r0wm8w r-fontFamily-vnw8o6 r-fontSize-ubezar r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
  dir="auto"
>
  Example Text
</div>
`;
